### PR TITLE
perf(p2p): Make ingesting packets not block on channel unmarshal or channel receive logic

### PR DESCRIPTION
**Pull request by ValarDragon**
Handles a significant part of #3199

This separates out the entire left hand side of into its own goroutine that won't block further reading:
image

That is unblocking us from reading/ingesting packets 73% of the CPU time on a very light period of Osmosis traffic, and likely far more real-time as much of this is mutex blocked! (Separate PR's will improve I/O usage, and other non-I/O terms)

We guarantee that every channel receives packets in-order from a peer.

Furthermore this will allow concurrency across channels for getting their own packets unmarshalled and sent to respective reactors. (Meaning a mempool tx gossip will now not block on a consensus packet receive, unless the per-channel unmarshal/recv buffer is full. Right now its parameterized to be low, but @evan-forbes has a PR that composes with this, for adding a buffer into every reactor, that should fix this from being remotely likely)

To me, I think the main questions for this PR being merged are:
- Are we fine w/ my API change of: `mconn._recover` to `mconn.PanicRecover` (publicly exporting it)? I do like where I ended up putting the panic recover. 
- Is it fine to put "setup channels fn" as a field element in peer? 
- Do we want any new behavior tested? I'll test this on mainnet osmosis. Due to matching the panic recovery path the old code, I don't think theres concern over that being handled differently.
- Any other tests we think are missing? Nothing comes to mind

I'll get this tested on mainnet osmosis soon, to help build confidence :)

---

#### PR checklist

- [ ] Tests written/updated - Some of the behavior here appears to never have been tested, so I we could introduce a test for sending an undecodable packet on a channel that does exist.
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

**Review by evan-forbes — COMMENTED** [review]
we should totally not block on marshaling, that make a ton of sense to me! However we might be able to improve on this further. 

I could definitely be missing something, but it looks like we are breaking the previous guarantee where messages are processed in the order that they were sent by the peer. In most cases, I don't think this should have any effect. There could be some weird difficult to debug edge cases however. 

If we measure the time required to unmarshal messages and queue them in an envelop buffer, we can see that we have an average of around .03ms and a std of around .15ms. Doing the unmarshaling in parallel would be a nice bonus, but how much time do we expect to cut off?

msg unmarshaling time


With the other envelope buffering approach, we can see that in some circumstances we see those envelope buffer reaching over 1000 and frequently over 250, so I think if we don't want to block here, we should increase to beyond 250.

one slight modification that could be interesting would be to combine it with the previous attempt, so instead of only passing the envelopes in a buffer, we stick the msgBytes in the buffer. 

by default, it could look like [this](https://github.com/celestiaorg/celestia-core/commit/8e8a0da6a161bdd38d6b6754a520584608f11b95), but the neat thing about this approach is that we can leave parallelism to the reactor. instead of just unmarshaling in parallel, in reactors where its perfectly safe (such as the mempool), we can do all processing steps in parallel.

**Comment by ValarDragon** [comment]
> I could definitely be missing something, but it looks like we are breaking the previous guarantee where messages are processed in the order that they were sent by the peer. In most cases, I don't think this should have any effect. There could be some weird difficult to debug edge cases however.

We aren't breaking this guarantee within a channel. We are only potentially breaking this guarantee across channels. Though your right this could cause complicated edge cases -- You may receive a vote for a block before the proposal msg, even though the peer sent both.

Thats a good point though, maybe we should do this per-reactor. 

> one slight modification that could be interesting would be to combine it with the previous attempt, so instead of only passing the envelopes in a buffer, we stick the msgBytes in the buffer.

I think thats an amazing idea. If we want to do this without breaking changes so we can ship faster, maybe the move is:
- Change my code to one thread per reactor
- Do logic where "If reactor supports async receive bytes, use that. Else do whats in this PR"

That way we can ship with no breakage, and always get the guarantee we need?

File: p2p/peer.go [hunk 1]
```diff
@@ -211,6 +224,10 @@ func (p *peer) OnStop() {
 	if err := p.mconn.Stop(); err != nil { // stop everything and close the conn
 		p.Logger.Debug("Error while stopping peer", "err", err)
 	}
+	for i := 0; i < len(p.channels); i++ {
+		p.recvListenersQuitChan <- struct{}{}
```
Comment (melekes) [inline p2p/peer.go:228]: Why do we need to write `len(channels)` messages? Isn't `close(chan)` won't result in all goroutines receiving a msg?

**Review by melekes — COMMENTED** [review]
Thanks @ValarDragon ❤️

**Review by cason — COMMENTED** [review]
I have conflicting feelings regarding this solution:

1. We should pay the cost of unmarshalling at some point, but is it fair do that channel by channel?
2. Aren't we artificially pretending that we are able to consume 256 messages per channel, while this is not true, clearly? How we impose back-pressure in this case?

Maybe the better solution here is quite simpler: have a thread for unmarshalling data, which is a costly part. Notice that in the original design were Reactors to do unmarshalling. Most of these performance penalties are result of unmarshalling/marshalling in the p2p layer, instead of in the Reactors.

**Pull request closed without merging**